packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
     IERC20 sellToken = sellCelo ? getGoldToken() : IERC20(stableToken);
     require(
       sellToken.transferFrom(msg.sender, address(this), sellAmount),
-      "Transfer of sell token failed"
+      "Transfer in of sell token failed"
     );
 
     // Record the proposal.
-    uint256 proposalId = exchangeProposalCount;
-    exchangeProposals[proposalId] = ExchangeProposal({
+    // Add 1 to the running proposal count and use as the proposalId. This is
+    // intentionally never 0 because LinkedList, which is used for keeping track
+    // of active proposal IDs, requires keys to be non-zero.
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
       exchanger: msg.sender,
       stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
       sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
       buyAmount: buyAmount,
       approvalTimestamp: 0, // initial value when not approved yet
-      state: ExchangeState.Proposed,
+      state: ExchangeProposalState.Proposed,
       sellCelo: sellCelo
     });
-    exchangeProposalCount = exchangeProposalCount.add(1);
+    // Push it into the array of active proposals.
+    activeProposalIds.push(exchangeProposalCount);
     // Even if stable tokens are being sold, the sellAmount emitted is the "value."
     emit ExchangeProposalCreated(
-      proposalId,
+      exchangeProposalCount,
       msg.sender,
       stableToken,
       sellAmount,
       buyAmount,
       sellCelo
     );
+    return exchangeProposalCount;
+  }
 
-    return proposalId;
+  /**
+   * @notice Approves an existing exchange proposal.
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Set the time the approval occurred and change the state.
+    proposal.approvalTimestamp = block.timestamp;
+    proposal.state = ExchangeProposalState.Approved;
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
```

Comment:
why is this `nonReentrant` if it's only calling core contracts in this function?

Reply:
The pattern I've seen so far throughout our contracts is if a function is state-modifying and involves any calls to an outside contract, we use `nonReentrant` just to be super safe. Apart from the gas cost involved with the `nonReentrant` modifier I don't see a big downside in having it, wdyt?

Reply:
yeah, only worried about gas usage, not a big deal actually.